### PR TITLE
feat: support orchestrator and proxysql for mysql 

### LINKED ISSUE #353: [Features] ProxySQL and Orchastrator and MySQL Replication Cluster
Support an Addon with three componentDefs:
1. proxysql https://proxysql.com/
2. orchestrator https://github.com/openark/orchestrator
3. mysql replication topology (with one primary, and multiple secondary mysql instances)

And we shall support following feature step by step:
1. create a cluster with three aforementioned components
1. support configuration of each compoent
2. support monitong of  each compont
3. TBC.

### PR DESCRIPTION
**Pull request by kubeJocker**
- fix #353

### CODE REVIEW

**Comment by kubeJocker** [comment]
HOW TO USE:

kbcli cluster create mysql {you_meta_mysql_name} --mode raftGroup

helm upgrade  orchestrator ./orchestrator --install 

helm upgrade  orc-cluster ./orchestrator-cluster --install  --set serviceReference.orchestrator.cluster= \
{you_meta_mysql_name}  --set serviceReference.orchestrator.namespace={you_meta_mysql_namespace}

helm upgrade  mysql-orc ./mysql-orc --install 

helm upgrade  mysql-orc-c ./mysql-orc-cluster --install --set serviceReference.orchestrator.namespace=\
{you_orchestrator_namespace} --set serviceReference.orchestrator.cluster=orc-cluster

File: addons/mysql/templates/componentdefinition-mysql-5.7.yaml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        numSymbols: 0
+        letterCase: MixedCases
+  roleProbe:
+    builtinHandler: mysql
+    periodSeconds: {{ .Values.roleProbe.periodSeconds }}
+    timeoutSeconds: {{ .Values.roleProbe.timeoutSeconds }}
+  runtime:
+    initContainers:
+      - image: {{ .Values.image.registry | default "docker.io" }}/mysql_audit_log:{{ .Values.image.tag }}
+        name: init-data
```
Comment (xuriwuyun) [inline addons/mysql/templates/componentdefinition-mysql-5.7.yaml:59]: the command is missing, has it been tested?

File: addons/mysql/templates/componentdefinition-mysql-8.0.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        numSymbols: 0
+        letterCase: MixedCases
+  lifecycleActions:
+    roleProbe:
+      builtinHandler: mysql
+      periodSeconds: {{ .Values.roleProbe.periodSeconds }}
+      timeoutSeconds: {{ .Values.roleProbe.timeoutSeconds }}
+  runtime:
+    initContainers:
+      - image: {{ .Values.image.registry | default "docker.io" }}/mysql_audit_log:8.0.33
```
Comment (xuriwuyun) [inline addons/mysql/templates/componentdefinition-mysql-8.0.yaml:59]: the same problem?

File: addons/mysql/templates/componentdefinition-mysql-5.7.yaml (hunk 1 above)
Comment (kubeJocker) [reply addons/mysql/templates/componentdefinition-mysql-5.7.yaml:59]: fixed by copying the command from clusterDefination

File: addons/mysql/templates/componentdefinition-mysql-8.0.yaml (hunk 2 above)
Comment (kubeJocker) [reply addons/mysql/templates/componentdefinition-mysql-8.0.yaml:59]: fixed the same way

File: addons/apecloud-mysql/templates/clusterversion-auditlog.yaml [hunk 3]
```diff
@@ -23,7 +23,7 @@ spec:
               - |
                 cp /usr/lib/mysql/plugin/ {{ .Values.mysqlConfigs.dataMountPath }}/plugin -r
             imagePullPolicy: Always
-            image: {{ .Values.image.registry | default "docker.io" }}/apecloud/mysql_audit_log:8.0.33
+            image: {{ .Values.image.registry | default "docker.io" }}/mysql_audit_log:8.0.33
```
Comment (shanshanying) [inline addons/apecloud-mysql/templates/clusterversion-auditlog.yaml:26]: why remove 'apecloud' repo ? if we set image.registry to our aliyun registry, the link is invalid.

File: addons/mysql/scripts/init-meta-database.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#!/bin/sh
+set -ex
+# 定义 MySQL 连接参数
```
Comment (shanshanying) [inline addons/mysql/scripts/init-meta-database.sh:3]: translate to en

File: addons/mysql/scripts/mysql-orchestrator-register.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
+#!/bin/sh
+set -ex
+
+# 定义 MySQL 连接参数
```
Comment (shanshanying) [inline addons/mysql/scripts/mysql-orchestrator-register.sh:4]: cn  -> en

File: addons/mysql/scripts/mysql-orchestrator-register.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    response=$(curl -s -o /dev/null -w "%{http_code}" $url)
+    if [ $response -eq 200 ]; then
+        echo "response success"
+        break
+    fi
+  done
+  echo "register $pod_name ($host_ip) to Orchestrator successful."
+}
+
+# 从环境变量中获取 Pod 名称列表和 IP 地址列表
```
Comment (shanshanying) [inline addons/mysql/scripts/mysql-orchestrator-register.sh:60]: dito, and fix them all

File: addons/mysql/scripts/mysql-orchestrator-register.sh [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  local instance_url="http://${endpoint}/api/instance/$host_ip/3306"
+
+  echo "register first mysql pod to orchestrator..."
+  instanceResponse=$(curl -s -o /dev/null -w "%{http_code}" $instance_url)
+    if [ $instanceResponse -eq 200 ]; then
+      echo "response success"
+    fi
+
+  while true; do
+    # 注册到 Orchestrator
```
Comment (shanshanying) [inline addons/mysql/scripts/mysql-orchestrator-register.sh:40]: dito

File: addons/mysql/values.yaml [hunk 8]
```diff
@@ -3,13 +3,14 @@
 # Declare variables to be passed into your templates.
 
 image:
-  registry: 
-  repository: mysql
+  registry:  registry.cn-hangzhou.aliyuncs.com/apecloud
```
Comment (shanshanying) [inline addons/mysql/values.yaml:6]: Following setting must be consistent for apecloud hosted images:

registry     : docker.io  or  registry.cn-hangzhou.aliyuncs.com
repository: apecloud/mysql-jq

File: addons/orchestrator-cluster/values.yaml [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# Default values for mysql-cluster.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+## @param version MySQL cluster version
+##
+version: mysql-8.0.33
```
Comment (shanshanying) [inline addons/orchestrator-cluster/values.yaml:7]: use orch version pls.

File: addons/orchestrator-cluster/values.yaml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+## @param mode MySQL cluster topology mode, standalone, replication or raftGroup
+##
```
Comment (shanshanying) [inline addons/orchestrator-cluster/values.yaml:10]: remove unused parameters here, and update these descriptions.

File: addons/orchestrator/values.yaml [hunk 11]
```diff
@@ -0,0 +1,97 @@
+# Default values for mysql.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+image:
+  registry:  registry.cn-hangzhou.aliyuncs.com/apecloud
```
Comment (shanshanying) [inline addons/orchestrator/values.yaml:6]: registry: docker.io or registry.cn-hangzhou.aliyuncs.com
repo: apecloud/...

Comment (shanshanying) [reply addons/orchestrator/values.yaml:6]: These values for mysql or orchestrator ?

File: addons/orchestrator-cluster/templates/cluster.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+apiVersion: apps.kubeblocks.io/v1alpha1
+kind: Cluster
+metadata:
+  name: {{ include "kblib.clusterName" . }}
+  labels: {{ include "kblib.clusterLabels" . | nindent 4 }}
+  annotations:
+    kubeblocks.io/enabled-node-port-svc: orchestrator,mysql,proxysql
```
Comment (shanshanying) [inline addons/orchestrator-cluster/templates/cluster.yaml:7]: why we should use node-port for all components?

File: addons/orchestrator-cluster/templates/cluster.yaml [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    kubeblocks.io/enabled-pod-ordinal-svc: orchestrator,mysql,proxysql
```
Comment (shanshanying) [inline addons/orchestrator-cluster/templates/cluster.yaml:8]: mysql and proxsql are not even specified in this cluster, is it necessary to specify them here?

File: addons/orchestrator-cluster/templates/meta-servicedescriptor.yaml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{{- if eq .Values.serviceReference.orchestrator.cluster ""}}
+apiVersion: apps.kubeblocks.io/v1alpha1
+kind: ServiceDescriptor
+metadata:
+  name: meta-service-descriptor
```
Comment (shanshanying) [inline addons/orchestrator-cluster/templates/meta-servicedescriptor.yaml:5]: add namespace

File: addons/mysql/templates/componentdefinition-proxysql-5.7.yaml [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+kind: ComponentDefinition
+metadata:
+  name: {{ include "proxysql.componentDefName" . }}-5.7
+  labels:
+      {{- include "mysql.labels" . | nindent 4 }}
+spec:
+  provider: kubeblocks
+  description: mysql proxy definition for Kubernetes
+  serviceKind: proxysql
+  serviceVersion: 8.0.0
```
Comment (shanshanying) [inline addons/mysql/templates/componentdefinition-proxysql-5.7.yaml:11]: serviceVersion 5.7

File: addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+kind: ComponentDefinition
+metadata:
+  name: {{ include "mysql.componentDefName" . }}-5.7.44-orc
+  labels:
+      {{- include "mysql.labels" . | nindent 4 }}
+spec:
+  provider: kubeblocks
+  description: mysql  compoent definition for Kubernetes
+  serviceKind: mysql
+  serviceVersion: 8.0.0
```
Comment (shanshanying) [inline addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml:11]: serviceVersion 5.7

File: addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml [hunk 17]
```diff
@@ -0,0 +1,206 @@
+apiVersion: apps.kubeblocks.io/v1alpha1
+kind: ComponentDefinition
+metadata:
+  name: {{ include "mysql.componentDefName" . }}-5.7.44-orc
+  labels:
+      {{- include "mysql.labels" . | nindent 4 }}
+spec:
+  provider: kubeblocks
+  description: mysql  compoent definition for Kubernetes
```
Comment (shanshanying) [inline addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml:9]: typo `compoent` -> `component`, check and fix them all.

File: addons/mysql/templates/componentdefinition-mysql-5.7.yaml [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+      serviceName: mysql
+      generatePodOrdinalService: true
+      spec:
+        ports:
+          - name: mysql
+            port: 3306
+            targetPort: mysql
+  configs:
+    - name: mysql-replication-config
+      templateRef: oracle-mysql5.7-config-template
```
Comment (shanshanying) [inline addons/mysql/templates/componentdefinition-mysql-5.7.yaml:31]: `oracle-mysql` is for demo purpose. pls use configconstraint for mysql pls. check the name and replace them all (and better to ref the object of the same major version, do not mix them up).

File: addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+        command:
+          - bash
+          - -c
+          - |
+            mv {{ .Values.dataMountPath }}/plugin/audit_log.so /usr/lib64/mysql/plugin/
+            rm -rf {{ .Values.dataMountPath }}/plugin
+            chown -R mysql:root {{ .Values.dataMountPath }}
+            REPORT_HOST=MYSQL_ORDINAL_HOST_${KB_POD_NAME##*-}
+            /scripts/docker-entrypoint-8.0.sh mysqld --server-id $(( ${KB_POD_NAME##*-} + 1)) --report_host ${!REPORT_HOST} \
+            --plugin-load-add=rpl_semi_sync_source=semisync_source.so \
```
Comment (shanshanying) [inline addons/mysql/templates/componentdefinition-mysql-5.7-orc.yaml:115]: sync_source/replica for mysql 8.0.x and i guess it should be sync_master/slave for mysql 5.7.x

File: addons/mysql-cluster/templates/cluster.yaml [hunk 20]
```diff
@@ -5,16 +5,53 @@ metadata:
...
   clusterVersionRef: {{ .Values.version }}
   terminationPolicy: {{ .Values.extra.terminationPolicy }}
   {{- include "kblib.affinity" . | indent 2 }}
   clusterDefinitionRef: mysql # ref clusterdefinition.name
   componentSpecs:
     - name: mysql
+      replicas: {{ .Values.replicas }}
       componentDefRef: mysql # ref clusterdefinition componentDefs.name
-      {{- include "kblib.componentMonitor" . | indent 6 }}
+      {{- $clusterversion := .Values.version }}
```
Comment (shanshanying) [inline addons/mysql-cluster/templates/cluster.yaml:19]: cv and componentDefRef fields are not used and we should remove them and test how it works when referring differnt cmpds.

File: addons/mysql/scripts/docker-entrypoint-5.7.sh [hunk 21]
```diff
@@ -454,10 +454,19 @@ _main() {
 			echo
 			mysql_note "MySQL init process done. Ready for start up."
 			echo
+			mysql_note "setup_master_slave"
+      setup_master_slave
 		else
 			mysql_socket_fix
 		fi
 	fi
+
```
[the remark this answers is not in the document]
Comment (kubeJocker) [reply addons/mysql/scripts/docker-entrypoint-5.7.sh:461]: done

**Review by shanshanying — APPROVED** [review]

**Pull request merged**